### PR TITLE
Msvc compile fixes

### PR DESCRIPTION
**Pull request by jimidobson**
Implemented fixes for various ambiguous method calls which caused compile errors in MSVC. These include:

Specifying the return type of bind::() calls
Changed some binds to lambdas for cleaner code
add 'cycles' as a capture for the aiming lambda in dump.cpp

### CODE REVIEW

**Comment by jimidobson** [comment]
Finally got those pesky formatting issues sorted!

**Comment by woutershep** [comment]
Instead of 'auto foo = (bar) baz;' why don't you use 'bar foo = baz;' ?

**Comment by jimidobson** [comment]
I left the format as it is unless I felt confident I could convert into a lambda - I assumed the use of auto was purposeful.

**Comment by Coolthulhu** [comment]
> I assumed the use of auto was purposeful

It's generally intended to avoid having to specify long typenames that aren't relevant. For example, if you want an iterator but don't want to write `map<std::string, some_struct>::iterator iter`.

If you're specifying the type anyway, `auto` only really makes sense if you need an explicit cast.

Another use case for `auto`s is when the type may change in the future, but will certainly retain some traits. For example, `int` to `long`, `map` to `unordered_map` or `iterator` to `reverse_iterator`. `auto` will work in all those cases.

I'll need an explicit list of what is to be avoided for MSVC to properly build the repo.
Does it hate all `auto x = std::bind`?

**Comment by jimidobson** [comment]
MSVC only had an issue with the `read_from_file` and `read_from_file_optional` methods and only because the second parameter, the callback, wasn't known at the point of compile. As a result, MSVC doesn't know which version of the `read_from_file` to use and throws the error.

It wasn't the `std::bind` that was the issue, it was that the `read_from_file` methods had multiple signatures and the `std::bind` didn't specify which of them it was returning at compile time. That's why specifying the return type via casting fixed the issues.

However, In these circumstances, it seems reasonable to remove the `auto` keywords and just specifically declare the reader type if there isn't any specific reason to keep them as auto. Doing it would render those lines of code a bit more readable and shorten them slightly, as well as removing some redundancy.

I'd be happy to do that if that seems reasonable to you guys?

**Comment by Coolthulhu** [comment]
The type of callback is known at compile time in all those cases, MSVC is just lazy (other compilers can figure it out just fine).

It would be better to change the `read_from_file` signature rather than uses of the function to avoid this happening in the future. Otherwise any new usage of `read_from_file` will most likely break the fix.

**Comment by jimidobson** [comment]
I'll change the signatures such that each one is named after the type of callback it expects:

`read_from_file` becomes `read_from_file_istream`

This way, its obvious which one you need to call and it avoids ambiguity.

**Comment by jimidobson** [comment]
@Coolthulhu I've made the modifications - let me know what you think

**Comment by Coolthulhu** [comment]
Now that the usage is no longer ambiguous, the variable extraction thing should be removed.

There should be a comment on why is the function not overloaded but instead split into 3 ones in the header file with the functions.

Also, pinging @BevapDin who wrote the file functions and may have a better suggestion on how to fix this.

**Comment by jimidobson** [comment]
Is this issue still open? I'm not able to do much while MSVC compilation is broken :(

**Comment by Coolthulhu** [comment]
I explained what needs to be done.

Now that function names are explicit, you no longer need the variable extraction and should inline them back into an argument where it is cleaner to do so.

**Comment by BevapDin** [comment]
> Now that the usage is no longer ambiguous, the variable extraction thing should be removed.

Agreed.

> There should be a comment on why is the function not overloaded but instead split into 3 ones in the header file with the functions.

Yes, would be nice.

> It would be better to change the  read_from_file  signature rather than uses of the function to avoid this happening in the future. Otherwise any new usage of  read_from_file  will most likely break the fix.

That's an excellent solution.

> Also, pinging @BevapDin who wrote the file functions and may have a better suggestion on how to fix this.

He hasn't.

**Comment by jimidobson** [comment]
I've removed the variable extraction where necessary; I left alone parts where doing so didn't make sense or would make the code more difficult to read.

File: src/worldfactory.cpp [hunk 1]
```diff
@@ -1454,12 +1454,17 @@ bool worldfactory::load_world_options(WORLDPTR &world)
...
         return true;
     }
 
     const auto legacy_path = world->world_path + "/" + FILENAMES["legacy_worldoptions"];
-    if( read_from_file_optional( legacy_path, std::bind( &WORLD::load_legacy_options, world, _1 ) ) ) {
+    auto legacy_options_reader = [world]( std::istream &is ) {
+        world->load_legacy_options( is );
+    };
+
+    if( read_from_file_optional_istream( legacy_path, legacy_options_reader ) ) {
```
Comment (Coolthulhu) [inline src/worldfactory.cpp]: Why not inline?

Comment (jimidobson) [inline src/worldfactory.cpp]: That method call is part of a conditional - I don't think nesting the lambda within that conditional will enhance readability. If you prefer it that way, I will change it.

Comment (jimidobson) [inline src/worldfactory.cpp]: @Coolthulhu Would you like me to change this code? If you'd prefer that inlined, I'd be happy to do it.

**Comment by nornagon** [comment]
I patched this in and still get errors telling me to enable `_SCL_SECURE_NO_WARNINGS` https://msdn.microsoft.com/en-us/library/aa985974.aspx

I'm not sure what the origin of the error is, because VS just tells me it's in "xutility", which is a VS standard library file. Adding `/D_SCL_SECURE_NO_WARNINGS` to the command line makes it go away though.

**Comment by jimidobson** [comment]
I had that issue as well, I too had to add that to fix the issue. I'm not sure where it stems from.

**Comment by Coolthulhu** [comment]
Can we get a guide on how to fix this error?

Someone will need to confirm this fix is working. That's the standard procedure here - one person writes, other person tests. The tester usually has merge rights, but I guess we could make an exception here if we have no one with VS2015 set up.

**Comment by jimidobson** [comment]
I looked around and it stems from a check introduced in MSVC 8.0 that flags up a bunch of calls to std library calls that *could* be used in an unsafe way. I don't think the warning is relevant to what Cataclysm is doing, but MSVC insists nonetheless.

See here:  http://stackoverflow.com/questions/1301277/c-boost-whats-the-cause-of-this-warning

Other than going through the code with a fine-toothed comb to pick out all instances of such calls, I don't think there is a fix that could be reasonably implemented other than using the preprocessor command to suppress them. 

On the topic of fixes though, I'd be happy to make a separate pull request that just includes the .vcproj with this preprocessor command added and I guess someone else with VS2015 could test it?

**Comment by jimidobson** [comment]
An update on _SCL_SECURE_NO_WARNINGS:

The exact lines of code as of the most recent master are:

map.cpp line 7529 (std::copy_n)
player_activity.cpp line 543 (std::equal)

Commenting these lines out in VS2015 removes the compilation error, so these seem to be what's causing it.

Of course, I don't see any reason why these lines of code are problematic.

On consideration, I don't think adding a separate pull request for this will be necessary as adding the preprocessor directive is still an MSVC specific fix. I can add the directive to the Cataclysm.vcxproj
file and add it as part of this pull request. What do you think @Coolthulhu ?

**Comment by Coolthulhu** [comment]
Should be in this PR. Being able to compile the project in VS2015 sounds like a good, objective goal.

**Comment by nornagon** [comment]
I've been fiddling around with NuGet packages and I've managed to make a build that fetches SDL and related packages from NuGet. This drastically simplifies the process of setting up the project on Windows, as there's no need to download and configure SDL2, SDL2_image, SDL2_ttf, SDL2_mixer, libintl, and Lua. Instead, NuGet automatically fetches the appropriate code from the NuGet repos.

There are a few remaining things that need to be ironed out before it's ready for a PR (outlined below), but if there's interest, I'd be happy to follow up on those issues. I'd love for the windows setup process to be as simple as "double-click the .sln and hit 'build'". I had never heard of NuGet before firing up VS2015 this week, though, so I might be missing something that a more experienced windows dev would warn about.

The outstanding issues are:
- There's no working NuGet package for SDL_mixer. I've built one locally, but haven't yet uploaded it.
- Compiling with /DLOCALIZATION gives a link error, even though libintl is installed via NuGet. I'm not sure what's causing this yet. Turning off localization works around it for now.
- The Lua package doesn't include the lua.exe binary necessary for generating the lua bindings (only the dll). This could potentially be fixed by adding a "bootstrap" executable to just wrap the dll and call the bindings generation script, or by making a NuGet package that included a lua.exe that could be referenced during prebuild (or possibly by some other method I haven't thought of).

**Comment by jimidobson** [comment]
@nornagon I'd definitely be interested in a way to simplify setting up the project dependencies. I've not heard of NuGet before either so I don't have any advice to offer on that topic.

**Comment by Coolthulhu** [comment]
Is there a foolproof guide to getting it to compile without Lua, localization or tiles?
That would be enough to consider starting supporting VS.

**Comment by mugling** [comment]
Closed as stalled

**Pull request closed without merging**